src/Workspaces/Core/MSBuild/MSBuild/MSBuildWorkspace_ProjectReload.cs

```diff
@@ -0,0 +1,309 @@
+            {
+                project = project.WithProjectReferences(newProjectInfo.ProjectReferences);
+            }
+
+            // changed metadata references
+            if (oldCheckSum.MetadataReferences.Checksum != newCheckSum.MetadataReferences.Checksum)
+            {
+                project = project.WithMetadataReferences(newProjectInfo.MetadataReferences);
+            }
+
+            // changed analyzer references
+            if (oldCheckSum.AnalyzerReferences.Checksum != newCheckSum.AnalyzerReferences.Checksum)
+            {
+                project = project.WithAnalyzerReferences(newProjectInfo.AnalyzerReferences);
+            }
+
+            // changed documents
+            if (oldCheckSum.Documents.Checksum != newCheckSum.Documents.Checksum)
+            {
+                project = await UpdateDocumentsAsync(
+                    project,
+                    newProjectInfo.Documents,
+                    project.State.DocumentStates.States,
+                    oldCheckSum.Documents,
+                    newCheckSum.Documents,
+                    (solution, documents) => solution.AddDocuments(documents),
+                    (solution, documentId) => solution.RemoveDocument(documentId)).ConfigureAwait(false);
+            }
+
+            // changed additional documents
+            if (oldCheckSum.AdditionalDocuments.Checksum != newCheckSum.AdditionalDocuments.Checksum)
+            {
+                project = await UpdateDocumentsAsync(
+                    project,
+                    newProjectInfo.AdditionalDocuments,
+                    project.State.AdditionalDocumentStates.States,
+                    oldCheckSum.AdditionalDocuments,
+                    newCheckSum.AdditionalDocuments,
+                    (solution, documents) => solution.AddAdditionalDocuments(documents),
+                    (solution, documentId) => solution.RemoveAdditionalDocument(documentId)).ConfigureAwait(false);
+            }
+
+            // changed analyzer config documents
+            if (oldCheckSum.AnalyzerConfigDocuments.Checksum != newCheckSum.AnalyzerConfigDocuments.Checksum)
+            {
+                project = await UpdateDocumentsAsync(
+                    project,
+                    newProjectInfo.AnalyzerConfigDocuments,
+                    project.State.AnalyzerConfigDocumentStates.States,
+                    oldCheckSum.AnalyzerConfigDocuments,
+                    newCheckSum.AnalyzerConfigDocuments,
+                    (solution, documents) => solution.AddAnalyzerConfigDocuments(documents),
+                    (solution, documentId) => solution.RemoveAnalyzerConfigDocument(documentId)).ConfigureAwait(false);
+            }
+
+            return project;
+        }
+
+        private static Project UpdateProjectAttributes(Project project, ProjectInfo.ProjectAttributes attributes)
+        {
+            // there is no API to change these once project is created
```

Comment:
Isn't this all copy/pasted from RemoteWorkspace.SolutionCreator.cs? How can we refactor to avoid the duplication?